pep8 fixes, license and start unittest (Pytest)

frennkie opened the pull request:
Started "pep8"-ing the modules.. more to come into this branch/PR

cvandeplas committed:
office - list ActiveX objects in metadata output

Conflicts:
	viper/modules/pdftools

Rafiot committed:
Initial refactoring for python3 support

Rafiot committed:
Make pdftools a submodule

Rafiot committed:
Update submodules paths

Rafiot committed:
Display related fixes for python3

Rafiot committed:
Fix python3 rendering (py2 compatible)

Rafiot committed:
Make python2 happy

Rafiot committed:
Update peepdf version

Rafiot committed:
More python2/python3 fixes

Rafiot committed:
[office] Cleanup and improvement of python3 support

Rafiot committed:
[config] Fix python3 support

Rafiot committed:
Improve python3 support, remove dependency on peepdf

Rafiot committed:
Fix SWF module

Rafiot committed:
Set lastline config in config file.

Rafiot committed:
Update Koodus to the new way to access the file content

Rafiot committed:
More Python3 fixes

Rafiot committed:
More python 2/3 fixes

Rafiot committed:
Remove print statements, cleanup

Rafiot committed:
Update MachO import

Rafiot commented on setup.py:
```diff
@@ -1,4 +1,11 @@
+# -*- coding: utf-8 -*-
+# This file is part of Viper - https://github.com/viper-framework/viper
+# See the file 'LICENSE' for copying permission.
+from __future__ import unicode_literals
```
Is it really helping? http://python-future.org/unicode_literals.html

frennkie replied:
I would feel a lot better about with the whole String/Unicode handling (while maintaining support for Python 2.7)  if we had tests that prove that stuff that was working are still doing the same thing on Python 3. See also my remark on absolute_import.

Rafiot commented on setup.py:
```diff
@@ -1,4 +1,11 @@
+# This file is part of Viper - https://github.com/viper-framework/viper
+# See the file 'LICENSE' for copying permission.
+from __future__ import unicode_literals
+from __future__ import absolute_import
```
I don't want to add any code that would support Python 2.6, and absolute_import seems to do that: http://python-future.org/futurize.html?highlight=absolute_import#stage-1-safe-fixes
Are you sure it is useful?

frennkie replied:
I worked on a project last year where I invested some time in checking these __future__ imports and from which I copied the code without re-checking. In that project I had to make sure it worked on Python 2.6 as I had a machine running RHEL 6. 

Sorry for not checking.. I'll look into this again and clean that stuff out.

Rafiot commented on setup.py:
```diff
@@ -1,4 +1,11 @@
+# See the file 'LICENSE' for copying permission.
+from __future__ import unicode_literals
+from __future__ import absolute_import
+from __future__ import print_function
```
Same here, are you sure it is useful? I prefer fixing the code to a python >=2.7 syntax than adding imports for somewhat broken code.

Rafiot commented on tox.ini:
```diff
@@ -4,22 +4,98 @@
 [tox]
 envlist =
-    pep8,
+    py34
```
Please add 3.6 here.

Rafiot commented on tox.ini:
```diff
@@ -4,22 +4,98 @@
+commands =
+    bash -c \'echo "exit" | python viper-cli\'
+
+[testenv:py34]
```
Also 3.6.

Rafiot commented on tox.ini:
```diff
@@ -4,22 +4,98 @@
-      --exclude=.build,.tox,.eggs,dist,docs,viper/modules \
-      --max-line-length=200 \
-      []
+# Flake8 Configuration
```
Do we need that big commented blob?

Rafiot commented on viper/modules/pehash/pehasher.py:
```diff
@@ -20,18 +22,19 @@
     HAVE_BITSTRING = False
 
-from viper.common.out import *
+# from viper.common.out import *  # not used
```
Just remove the unused imports.

Rafiot commented on viper/modules/pehash/pehasher.py:
```diff
@@ -20,18 +22,19 @@
 
 def calculate_pehash(file_path=None, data=None):
     if not HAVE_PEFILE:
+        # TODO (frennkie): what's going on here? should this be a Class (sub classing "Module")?
```
You're right, this is going to fail... Need to check how this module is supposed to work.

Rafiot commented on viper/modules/rats/blackshades.py:
```diff
@@ -28,20 +35,23 @@ def decrypt_configuration(hex):
     for x in range(0xffffff):
         PRNG_SEED = x
-        if get_next_rng_value() != pre_check[0] or get_next_rng_value() != pre_check[1] or get_next_rng_value() != pre_check[2]:
+        if get_next_rng_value() != pre_check[0] or get_next_rng_value() != pre_check[1] or get_next_rng_value() != \
```
This one can stay where it is, or cut the line before.

Rafiot commented on viper/modules/rats/jrat.py:
```diff
@@ -158,13 +169,14 @@ def parse_config(raw_config, enckey):
 def config(data):
     enckey, conf = get_parts(data)
-    if enckey == None:
+    if not enckey:
```
Did you make sure of the comportment if enckey is 0, '' or []? `if enckey is None` might be preferred there.

frennkie replied:
This was actually auto suggest by the pep8 fixes. Also looks good on both py27 and py34:

```
>>> if 0 or "" or []:
...     print("something there")
... else:
...     print("nope")
...
nope
```

Rafiot replied:
Sure, `0, "" and []` are `False` but are we supposed to return if we have an empty list?

frennkie replied:
Not quite sure what you are getting at. This change is purely syntax.. it does not change the behavior. 

Of course you can say "if `enckey == None` is exactly the same as `if not enckey` - why change it.. it's a question about style and convention.

Rafiot replied:
My point is that isn't the same ;) 

`if enckey is None` is the same as `if enckey == None`, when `if not enckey` is the same as `if enckey == 0 or len(enckey) == 0 or enckey is None`

frennkie replied:
Yeah.. you are right (again ;-) ) I changed is to `if enckey is None`.

Rafiot commented on viper/modules/virustotal.py:
```diff
@@ -355,11 +368,17 @@ def run(self):
                             self.log('success', 'You opened a sample related to the current MISP event.')
                     else:
-                        self.log('warning', 'This samples is linked to the MISP event {eid}. You may want to run misp pull {eid}'.format(eid=eid))
+                        self.log('warning',
```
Hmmm this is seriously painful to read. Please put those ones back on one line.

frennkie replied:
How do you feel about a max character per line line limit in general? I know there are laods and loads of discussion (maybe even flamewars) on this topic. 

My opinion is that having a limit (e.g. 100/120) is useful - but of course exceptions can we easily be made on a per file or per line basis.

Rafiot replied:
Yes, I'm cool with 100/120, but only f it doesn't get in the way of reading the code.

Rafiot reviewed (commented):
I haz coments :) 

The one on the imports applies to all.

frennkie committed:
pep8 fixes, add license - commit squash

frennkie committed:
fix typos

frennkie committed:
Update .travis.yml

file became invalid through comments.. fix this

frennkie commented on .travis.yml:
```diff
@@ -21,7 +21,11 @@ before_install:
+    - git config user.name "Travis"
+    - if [[ -d viper/modules/pdftools ]]; then rm -rf viper/modules/pdftools ; fi
+    - git pull --recurse-submodules --no-commit && git submodule update
+
```
I had to add these four lines to get Travis to run.. should you accept to merge this, then you can clean this up...

frennkie committed:
clean up: remove blank lines

frennkie committed:
add pytest and some first tests

frennkie committed:
minor pep8 fixes

frennkie committed:
rename test file

frennkie committed:
undo change to .gitmodules

frennkie committed:
fix submodule pdftools

frennkie committed:
adapt travis cfg for possible fix

frennkie committed:
replace tabs with spaces

frennkie committed:
sperate submodule commands

frennkie commented:
@Rafiot I noticed that a PR which is merged in master (of viper-framework/viper) is missing from your py3 branch (from which I am currently working): https://github.com/viper-framework/viper/pull/481

Is there a reason for this? Are you waiting with rebasing until later?

Rafiot commented on viper/modules/pehash/pehasher.py:
```diff
@@ -6,30 +6,11 @@
 import bz2
 import hashlib
-
-try:
```
Having try-except clauses all over the code is expected, as not al the users want to install *all* the dependencies.

frennkie replied:
So if people don't install the dependency they won't have access to the info/analysis, but Viper won't crash.. good point.

Rafiot replied:
yep, most of the modules do that, so we should stick to that behaviour.

frennkie replied:
@Rafiot I recall now why I changed this: 
 * `calculate_pehash` is only ever called from pe.py - which itself already checks whether `pehash` [is availble](https://github.com/Rafiot/viper/blob/production/viper/modules/pe.py#L922) and prints a dependency message for the user if not.
* pehasher.py uses `self.log` without being a Class or having `self.log` - thus it would fail/crash if this code would be reachable in any way.

Rafiot commented:
I only merged in my py3 branch the changes that were creating conflicts. This one wasn't so I didn't merge it back in my branch. 

When done, I'll merge py3 into master so it shouldn't matter.

Rafiot commented:
I'm not into having the binary test files in the repository. Can you make it an external repo, please? 

If we have lots of test cases, it will increase the size of the main repo *a lot*

frennkie commented:
> If we have lots of test cases, it will increase the size of the main repo a lot

You mean similar to the [pefile repo](https://github.com/erocarrera/pefile/tree/master/tests)?! ;-)

I agree.. this is not nice. The tests I started are using Pytest which has a plugin that can work with local files: [pytest-datafiles](https://pypi.python.org/pypi/pytest-datafiles). In my opinion this would be a nice way to provide multiple test files without having them into this repo (just pull/download them to a local path and use that as a pytest fixture).

frennkie commented:
Just a random thought: If people (that is users) install Viper via pip from Pypi then having large files in the GH repo wouldn't really be a big issue.. Because they wouldn't be included in the package.. And a dev pulling from GH would need the binary files to run the full test suite..

But in the end I don't mind separating the files.

Rafiot commented:
Yes, such as the pefile repo :) 

Many viper user will clone the repo directly, just because this way they use HEAD, when pefile is more a library that people install with pip. So yeah, if we have an alternative to that, that would be great :)

frennkie commented:
I moved the files to [viper-test-file](https://github.com/frennkie/viper-test-files). The repo is expected to be available at `tests/viper-test-files` (~~I haven't add this as submodule (yet) - but this can of course be~~ done).

frennkie commented:
@Rafiot could you please check out these two commit:

* https://github.com/frennkie/viper/commit/17d27f475dfe566cb9d041260ce960292ebcd169 - should be obvious, right?!
* https://github.com/frennkie/viper/commit/6882926190fd9800abea80ed2cc34446e1d6e99b - I created a test for a randomly chosen RAT module and it failed when access File.data with StringIO. BytesIO fixes it for my sample. Is this a general issue?

Rafiot commented:
Not sure what you mean by "check out". Do you want me to cherry pick them now? 

* https://github.com/frennkie/viper/commit/17d27f475dfe566cb9d041260ce960292ebcd169: yes, sure, makes sense.
* https://github.com/frennkie/viper/commit/6882926190fd9800abea80ed2cc34446e1d6e99b: BytesIO will give you a byte stream (instead of a string). And should be used every time the file opened isn't plain text. In that case, yes, StringIO was a bug.

frennkie commented:
More like "look into". Sorry for the ambiguity.. ;-D

frennkie commented:
But there is the question how we should proceed with this pull request which became pretty extensive..  I also feel a little bit of a "feature/scope creep".

Do you still have questions/objections or improvement suggestions? If not I would appreciate if you could merge this.

For any future work I think I would the open new PRs which would then be more focused..  What do you think?

Rafiot commented:
I'll try to have a closer look at it later today, or tomorrow. But before merging, I'll definitely ask you to squash the commits: the commit history contains all the binaries, even if you deleted them in a later commit ;) 

And yes, smaller PR are better, no doubt :)

frennkie commented:
> But before merging, I'll definitely ask you to squash the commits

Before I squashed there were no conflicts.. now there are loads.. I may have squashed some commits that were actually not from me.. :-/

frennkie committed:
add pep8 and py3 changes and checks - squashed

frennkie commented:
I did not manage to squash all commits.. (without having too many conflicts) but it looks cleaner now and the binary files should be purged from the history (at least I couldn't find any traces).

frennkie committed:
move test files to submodule

frennkie committed:
reactivate pdftools

Rafiot commented:
Thank you very much, I'll look at it one more time asap

frennkie commented on viper/modules/rats/xrat.py:
```diff
@@ -81,22 +90,23 @@ def get_parts(long_line):
 def config(data):
     long_line, ver = get_long_line(data)
-    if ver == None:
+    if ver:
```
Whoops.. that's wrong

frennkie committed:
fix incorrect is (not) None fixes

Rafiot commented on viper/modules/office.py:
```diff
@@ -292,8 +293,6 @@ def xmlmeta(self, zip_xml):
                 embedded_list.append(name.split('/')[-1])
             if name == 'word/vbaProject.bin':
                 vba_list.append(name.split('/')[-1])
-            if name.startswith('word/activeX/'):
```
Why is it removed?

frennkie replied:
I believe I'm missing this commit: https://github.com/Rafiot/viper/commit/4f8df20d0e069cf4194e1efae3ec0d0aaae0dee9 not sure how that happened...

Rafiot commented on viper/modules/office.py:
```diff
@@ -308,10 +307,6 @@ def xmlmeta(self, zip_xml):
             self.log('info', "Macro Objects")
             for item in vba_list:
                 self.log('item', item)
-        if len(activex_list) > 0:
```
Same here

Rafiot commented on viper/modules/pe.py:
```diff
@@ -254,7 +258,8 @@ def get_compiletime(pe):
 
-        compile_time = get_compiletime(self.pe)
+        self.result_compile_time = get_compiletime(self.pe)
+        compile_time = self.result_compile_time
```
Why this variable instead of using `self.result_compile_time`

frennkie replied:
The reason why I introduced "self.result_compile_time" is that I wanted to access the "raw" result of calling pe.PE().run() (with argmument "compiletime"). Parsing the output from stdout is possible but not really nice. I also stumbled over the fact that the compile_time is printed in **bold**. 

In my opinion it would be useful to have a general concept of how the results of the modules can and should be accessed. I'm thinking along the lines of adding a `self.result = None` (or ` self.result = {}`) to the Module class. Then when an instance runs any code it should store the result in that variable/dict. This would be useful for testing and also for the API/Web interface.

But this is not really related/tied to the Python3 effort.. so should I open this as a Issue/PR on the Viper master?

Rafiot replied:
okay, makes sense.

Rafiot commented on viper/modules/pehash/pehasher.py:
```diff
@@ -1,37 +1,16 @@
+import pefile
+import bitstring
 
-try:
```
Please keep the try-except here.

frennkie replied:
I'm ok with leaving this in.. but the reason why I removed it here is that there are only two occurences where `calculate_pehash` is used: [pe.py#L938](https://github.com/Rafiot/viper/blob/py3/viper/modules/pe.py#L938) and [pe.py#L948](https://github.com/Rafiot/viper/blob/py3/viper/modules/pe.py#L948). At this point pe.py has already checked whether HAVE_PEHASH is true... (and raised the dependency message is not)

Rafiot replied:
Okay, no need to change it then, as long as it doesn't raise an exception, I'm cool with it.

Rafiot commented on viper/modules/rats/hawkeye.py:
```diff
@@ -1,33 +1,36 @@
-        return value#.encode('hex')
-    #except:
-        #return False
+    # try:
```
Please remove the commented try-except

Rafiot commented on viper/modules/rats/luxnet.py:
```diff
@@ -1,49 +1,55 @@
-        return unpack(">H", byteStr)[0]
-    #except:
-        #return None
+    # try:
```
Please remove the commented try-except

Rafiot commented:
Looks great!

frennkie committed:
clean up

Rafiot commented:
@frennkie Can you enable the tests in the travis file? 

Besides that, I think this PR is ready to be merged.

frennkie commented:
On startup I'm getting the warning message `[!] Something wrong happened while importing the module viper.modules.pymacho_helper.readmacho: No module named 'pymacho'` for the file [readmacho.py](https://github.com/viper-framework/Mach-O/blob/3340204da980a35b7988bb7bbffacd01092a0ad6/readmacho.py)  . As far as I can tell the file is not necessary for Viper.. so @Rafiot - can it be removed?

frennkie commented:
> Can you enable the tests in the travis file?

@Rafiot sure.. but I haven't set up any projects with Travis yet - so I'm not sure what the best way to do it is.

The pep8 checks are invoked by doing either `flake8 .` or by running `tox -e pep8`. I have not yet fixed apk.py but all other files have no remaining issues (a few files e.g. docs/source/conf.py are excluded).

The unittest can be called like this 
* `python setup.py test`
* `pytest`
* `tox` (to run all default environments)
* `tox -e py27,py36` (e.g. to run specific environments)

I would suggest to simply add `tox` to the .travis.yml configuration and Travis will then run everything that is setup in the tox.ini. I think it is useful to have the option to run the full test suite locally without having to commit and push.

Rafiot commented:
Good catch with the macho module, I removed the file from the automatic import.

I think I'll do a last review of the changes today and merge, and when it is merged, update the travis file. And yes, it definitely make sense to be able to run the tests locally, it is also good to have it automatically run on every commit/PR

frennkie committed:
fix broken androguard imports

frennkie commented:
I completely agree with you. 

I now updated the apk.py file - which now appears to work with [androguard master](https://github.com/androguard/androguard/commits/master). From my perspective that concludes the PEP8 fixes. But of the few tests I added already a few need to be checked and of course to be really sure that the Python3 enablement doesn't break anything a lot more tests are needed.

Rafiot commented:
Thank you very much for that! 

Final lookup later today, and I'll merge :)
